File: bindings/swift/examples/PackageInspect.swift
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// This example inspects a published Move package on testnet and prints its
// upgrade policy, version history, dependencies, functions, types, and sample
// objects.

import Foundation
import IotaSDK

private let frameworkPackageId = Address.framework().toHex()

@main
struct PackageInspectExample {
  static func main() async throws {
    let packageId = "0x6f727ea576a00036657fff0ae3a6d7c8171b178bf35112d6b83b2a6272cc5f0d"

    let packageAddress = try Address.fromHex(hex: packageId)
    let client = GraphQlClient.newTestnet()

    // Fetch package metadata and version history.
    guard let package = try await client.package(address: packageAddress) else {
      throw NSError(
        domain: "PackageInspect", code: 2,
        userInfo: [NSLocalizedDescriptionKey: "missing package"])
    }
    guard let latestPackage = try await client.packageLatest(address: packageAddress) else {
      throw NSError(
        domain: "PackageInspect", code: 3,
        userInfo: [NSLocalizedDescriptionKey: "missing latest package"])
    }

    let versions = try await fetchPackageVersions(client: client, packageAddress: packageAddress)
    let packagePrefix = package.id().toHex()
    print("Latest version: \(latestPackage.version().asU64()) (\(latestPackage.id().toHex()))")
    // Resolve the current upgrade policy.
    let currentPolicy = try await currentPackagePolicy(client: client, packageId: package.id())
    print("Current package policy: \(currentPolicy)")
    print()

    // Print the package version history.
    print("Versions:")
    for version in versions {
      var labels: [String] = []
      if version.id() == package.id() {
        labels.append("requested")
      }
      if version.id() == latestPackage.id() {
        labels.append("latest")
      }

      let suffix = labels.isEmpty ? "" : " [\(labels.joined(separator: ", "))]"
      print("- v\(version.version().asU64()) -> \(version.id().toHex())\(suffix)")
    }
    print()

    // Print package dependencies and their linked versions.
    print("Dependencies:")
    let dependencies = package.linkageTable().values.sorted {
      $0.upgradedId.toHex() < $1.upgradedId.toHex()
    }
    if dependencies.isEmpty {
      print("- none")
    } else {
      for upgrade in dependencies {
        print("- \(upgrade.upgradedId.toHex()) @ v\(upgrade.upgradedVersion.asU64())")
      }
    }
    print()

    // Inspect normalized modules, functions, types, and sample key objects.
    print("Package contents:")
    let moduleNames = package.modules().keys.map { $0.asStr() }.sorted()
    for moduleName in moduleNames {
      print("Module: \(moduleName)")

      let module = try await client.normalizedMoveModule(
        package: packageAddress,
        module: moduleName,
        version: nil,
        paginationFilterEnums: forwardPage(),
        paginationFilterFriends: forwardPage(),
        paginationFilterFunctions: forwardPage(),
        paginationFilterStructs: forwardPage()
      )

      guard let module else {
        print("  metadata: missing")
        print()
        continue
      }

      if let functions = module.functions, !functions.nodes.isEmpty {
        print("  functions:")
        for function in functions.nodes {
          let signature = formatFunctionSignature(
            String(describing: function),
            packagePrefix: packagePrefix
          )
          print("    - \(signature)")
        }
        if functions.pageInfo.hasNextPage {
          print("    - ...")
        }
      } else {
        print("  functions: none")
      }

      if let structs = module.structs, !structs.nodes.isEmpty {
        print("  types:")
        for structType in structs.nodes {
          let typeTag = "\(packagePrefix)::\(moduleName)::\(structType.name)"
          print("    - \(typeTag)")
          let hasKeyAbility = (structType.abilities ?? []).contains(.key)
          let isGeneric = !(structType.typeParameters ?? []).isEmpty
          try await printObjectSamples(
            client: client,
            typeTag: typeTag,
            hasKeyAbility: hasKeyAbility,
            isGeneric: isGeneric
          )
        }
        if structs.pageInfo.hasNextPage {
          print("    - ...")
        }
      } else {
        print("  types: none")
      }

      print()
    }
  }
}

private func forwardPage(cursor: String? = nil) -> PaginationFilter {
  PaginationFilter(direction: .forward, cursor: cursor)
}

private func isHexDigit(_ character: Character) -> Bool {
  character.unicodeScalars.allSatisfy { scalar in
    switch scalar.value {
    case 48...57, 65...70, 97...102:
      return true
    default:
      return false
    }
  }
}

private func shortenPackageIds(_ signature: String) -> String {
  var shortened = ""
  shortened.reserveCapacity(signature.count)

  var index = signature.startIndex
  while index < signature.endIndex {
    let nextIndex = signature.index(after: index)
    if signature[index] == "0", nextIndex < signature.endIndex, signature[nextIndex] == "x" {
      var end = signature.index(after: nextIndex)
      while end < signature.endIndex, isHexDigit(signature[end]) {
        end = signature.index(after: end)
      }

      if end > signature.index(after: nextIndex) {
        let candidate = String(signature[index..<end])
        if let address = try? Address.fromHex(hex: candidate) {
          shortened += address.toShortHex()
        } else {
          shortened += candidate
        }
        index = end
        continue
      }
    }

    shortened.append(signature[index])
    index = nextIndex
  }

  return shortened
}

private func formatFunctionSignature(_ signature: String, packagePrefix: String) -> String {
  shortenPackageIds(signature.replacingOccurrences(of: "\(packagePrefix)::", with: ""))
}

private func fetchPackageVersions(client: GraphQlClient, packageAddress: Address) async throws
  -> [MovePackage]
{
  var versions: [MovePackage] = []
  var cursor: String?

  while true {
    let page = try await client.packageVersions(
      address: packageAddress,
      paginationFilter: forwardPage(cursor: cursor)
    )
    versions.append(contentsOf: page.data)
    if page.pageInfo.hasNextPage {
      cursor = page.pageInfo.endCursor
    } else {
      break
    }
  }

  return versions.sorted { $0.version().asU64() < $1.version().asU64() }
}

private func printObjectSamples(
  client: GraphQlClient,
  typeTag: String,
  hasKeyAbility: Bool,
  isGeneric: Bool
) async throws {
  if !hasKeyAbility {
    return
  }

  if isGeneric {
    print("    sample objects: skipped for generic type")
    return
  }

  let objects = try await client.objects(
    filter: ObjectFilter(typeTag: typeTag),
    paginationFilter: PaginationFilter(direction: .forward, limit: 3)
  )

  if objects.data.isEmpty {
    print("    sample objects: none found")
    return
  }

  print("    sample objects:")
  for object in objects.data {
    print("      - \(object.id().toHex()) (version \(object.version().asU64()))")
  }
  if objects.pageInfo.hasNextPage {
    print("      - ...")
  }
}

private func formatPolicyName(_ policy: UInt8) -> String {
  switch policy {
  case 0:
    return "Compatible"
  case 128:
    return "Additive"
  case 192:
    return "Dependency-only"
  default:
    return "Unknown (\(policy))"
  }
}

private func extractPolicy(contents: Value) -> UInt8? {
  guard
    let data = contents.data(using: .utf8),
    let rawJson = try? JSONSerialization.jsonObject(with: data),
    let json = rawJson as? [String: Any],
    let rawPolicy = json["policy"]
  else {
    return nil
  }

  if let number = rawPolicy as? NSNumber {
    return number.uint8Value
  }
  if let text = rawPolicy as? String {
    return UInt8(text)
  }

  return nil
}

private func resolveUpgradeCapId(
  client: GraphQlClient,
  packageId: ObjectId
) async throws -> ObjectId? {
  let page = try await client.transactionsEffects(
    filter: TransactionsFilter(changedObject: packageId),
    paginationFilter: PaginationFilter(direction: .forward, limit: 1)
  )

  for effects in page.data {
    let effectsV1 = effects.asV1()
    for changedObj in effectsV1.changedObjects() {
      guard case .objectWrite = changedObj.outputState else {
        continue
      }

      if let object = try await client.object(
        objectId: changedObj.objectId,
        version: effectsV1.lamportVersion()
      ),
        object.asOptStruct()?.structType == StructTag.newUpgradeCap()
      {
        return changedObj.objectId
      }
    }
  }

  return nil
}

private func sameObjectId(_ left: String?, _ right: String?) -> Bool {
  guard let left, let right else {
    return false
  }

  return left.caseInsensitiveCompare(right) == .orderedSame
}

private func programmableTransactionJson(_ tx: Transaction) throws -> [String: Any]? {
  guard
    let data = try tx.toJson().data(using: .utf8),
    let rawJson = try JSONSerialization.jsonObject(with: data) as? [String: Any],
    let txV1 = rawJson["1"] as? [String: Any],
    let kind = txV1["kind"] as? [String: Any],
    let kindName = kind["kind"] as? String,
    kindName == "programmable_transaction"
  else {
    return nil
  }

  return kind
}

private func isPackageMakeImmutableCall(_ command: [String: Any]) -> Bool {
  sameObjectId(command["package"] as? String, frameworkPackageId)
    && command["command"] as? String == "move_call"
    && command["module"] as? String == "package"
    && command["function"] as? String == "make_immutable"
}

private func inputMatchesObjectId(_ input: [String: Any], objectId: String) -> Bool {
  guard
    let inputType = input["type"] as? String,
    ["immutable_or_owned", "receiving", "shared"].contains(inputType)
  else {
    return false
  }

  return sameObjectId(input["object_id"] as? String, objectId)
}

private func publishesPackageAsImmutable(_ tx: Transaction) throws -> Bool {
  guard
    let programmableTx = try programmableTransactionJson(tx),
    let commands = programmableTx["commands"] as? [[String: Any]]
  else {
    return false
  }

  let publishIndexes = commands.enumerated().compactMap { index, command in
    command["command"] as? String == "publish" ? index : nil
  }
  guard publishIndexes.count == 1, let publishIndex = publishIndexes.first else {
    return false
  }

  for command in commands.dropFirst(publishIndex + 1) {
    guard
      isPackageMakeImmutableCall(command),
      let arguments = command["arguments"] as? [Any],
      arguments.count == 1,
      let argument = arguments[0] as? [String: Any],
      let result = argument["result"] as? NSNumber,
      result.intValue == publishIndex
    else {
      continue
    }

    return true
  }

  return false
}

private func usesUpgradeCapForMakeImmutable(
  _ tx: Transaction,
  upgradeCapId: ObjectId
) throws -> Bool {
  guard
    let programmableTx = try programmableTransactionJson(tx),
    let inputs = programmableTx["inputs"] as? [[String: Any]],
    let commands = programmableTx["commands"] as? [[String: Any]]
  else {
    return false
  }

  let upgradeCapInputs = inputs.enumerated().compactMap { index, input in
    inputMatchesObjectId(input, objectId: upgradeCapId.toHex()) ? index : nil
  }
  guard !upgradeCapInputs.isEmpty else {
    return false
  }

  for command in commands {
    guard
      isPackageMakeImmutableCall(command),
      let arguments = command["arguments"] as? [Any],
      arguments.count == 1,
      let argument = arguments[0] as? [String: Any],
      let input = argument["input"] as? NSNumber,
      upgradeCapInputs.contains(input.intValue)
    else {
      continue
    }

    return true
  }

  return false
}

private func wasPackagePublishedAsImmutable(
  client: GraphQlClient,
  packageId: ObjectId
) async throws -> Bool {
  var cursor: String?

  while true {
    let page = try await client.transactionsDataEffects(
      filter: TransactionsFilter(changedObject: packageId),
      paginationFilter: forwardPage(cursor: cursor)
    )

    for txData in page.data {
      if try publishesPackageAsImmutable(txData.tx.transaction) {
        return true
      }
    }

    if !page.pageInfo.hasNextPage {
      return false
    }

    cursor = page.pageInfo.endCursor
  }
}

private func wasUpgradeCapUsedForMakeImmutable(
  client: GraphQlClient,
  upgradeCapId: ObjectId
) async throws -> Bool {
  var cursor: String?

  while true {
    let page = try await client.transactionsDataEffects(
      filter: TransactionsFilter(inputObject: upgradeCapId),
      paginationFilter: forwardPage(cursor: cursor)
    )

    for txData in page.data {
      if try usesUpgradeCapForMakeImmutable(
        txData.tx.transaction,
        upgradeCapId: upgradeCapId
      ) {
        return true
      }
    }

    if !page.pageInfo.hasNextPage {
      return false
    }

    cursor = page.pageInfo.endCursor
  }
}

private func currentPackagePolicy(
  client: GraphQlClient,
  packageId: ObjectId
) async throws -> String {
  guard let upgradeCapId = try await resolveUpgradeCapId(client: client, packageId: packageId)
  else {
    if try await wasPackagePublishedAsImmutable(client: client, packageId: packageId) {
      return "Immutable"
    }
    return "Unavailable"
  }

  guard let contents = try await client.moveObjectContents(objectId: upgradeCapId) else {
    if try await wasUpgradeCapUsedForMakeImmutable(client: client, upgradeCapId: upgradeCapId) {
      return "Immutable"
    }
    return "Unavailable"
  }

  return extractPolicy(contents: contents).map(formatPolicyName) ?? "Unavailable"
}

File: bindings/wasm/examples/package_inspect.mjs
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
//
// This example inspects a published Move package on testnet and prints its
// upgrade policy, version history, dependencies, functions, types, and sample
// objects.

import {
  Address,
  Direction,
  GraphQlClient,
  MoveAbility,
  ObjectFilter,
  PaginationFilter,
  StructTag,
  TransactionsFilter,
  initAsync,
} from "@iota/sdk-wasm";

await initAsync();

const FRAMEWORK_PACKAGE_ID = Address.framework().toHex();
const HEX_DIGITS = new Set("0123456789abcdefABCDEF");

function forwardPage(cursor = undefined) {
  return PaginationFilter.new({ direction: Direction.Forward, cursor });
}

function shortenPackageIds(signature) {
  const parts = [];
  let index = 0;
  while (index < signature.length) {
    if (signature.startsWith("0x", index)) {
      let end = index + 2;
      while (end < signature.length && HEX_DIGITS.has(signature[end])) {
        end += 1;
      }
      if (end > index + 2) {
        const candidate = signature.slice(index, end);
        try {
          parts.push(Address.fromHex(candidate).toShortHex());
          index = end;
          continue;
        } catch {
          parts.push(candidate);
          index = end;
          continue;
        }
      }
    }
    parts.push(signature[index]);
    index += 1;
  }
  return parts.join("");
}

function formatFunctionSignature(signature, packagePrefix) {
  return shortenPackageIds(signature.replaceAll(`${packagePrefix}::`, ""));
}

async function fetchPackageVersions(client, packageAddress) {
  const versions = [];
  let cursor = undefined;
  while (true) {
    const page = await client.packageVersions(
      packageAddress,
      undefined,
      undefined,
      forwardPage(cursor),
    );
    versions.push(...page.data);
    if (page.pageInfo.hasNextPage) {
      cursor = page.pageInfo.endCursor;
    } else {
      break;
    }
  }
  versions.sort((a, b) => Number(a.version().asU64() - b.version().asU64()));
  return versions;
}

async function printObjectSamples(client, typeTag, hasKeyAbility, isGeneric) {
  if (!hasKeyAbility) return;
  if (isGeneric) {
    console.log("    sample objects: skipped for generic type");
    return;
  }
  const objects = await client.objects(
    ObjectFilter.new({ typeTag }),
    PaginationFilter.new({ direction: Direction.Forward, limit: 3 }),
  );
  if (objects.data.length === 0) {
    console.log("    sample objects: none found");
    return;
  }
  console.log("    sample objects:");
  for (const obj of objects.data) {
    console.log(
      `      - ${obj.id().toHex()} (version ${obj.version().asU64()})`,
    );
  }
  if (objects.pageInfo.hasNextPage) console.log("      - ...");
}

function formatPolicyName(policy) {
  return (
    { 0: "Compatible", 128: "Additive", 192: "Dependency-only" }[policy] ??
    `Unknown (${policy})`
  );
}

function extractPolicy(contents) {
  let parsed;
  try {
    parsed = JSON.parse(contents);
  } catch {
    return null;
  }
  const policy = parsed?.policy;
  if (typeof policy === "string")
    return /^\d+$/.test(policy) ? Number(policy) : null;
  if (typeof policy === "number") return policy;
  return null;
}

async function resolveUpgradeCapId(client, packageId) {
  const page = await client.transactionsEffects(
    TransactionsFilter.new({ changedObject: packageId }),
    PaginationFilter.new({ direction: Direction.Forward, limit: 1 }),
  );
  for (const effects of page.data) {
    const effectsV1 = effects.asV1();
    for (const changedObj of effectsV1.changedObjects()) {
      if (!changedObj.outputState.isObjectWrite()) continue;
      const obj = await client.object(
        changedObj.objectId,
        effectsV1.lamportVersion(),
      );
      if (obj !== null && obj.asStructOpt() !== null) {
        if (
          obj.asStruct().structType.eq?.(StructTag.newUpgradeCap()) ??
          false
        ) {
          return changedObj.objectId;
        }
      }
    }
  }
  return null;
}

function sameObjectId(left, right) {
  return typeof left === "string" && left.toLowerCase() === right.toLowerCase();
}

function programmableTransactionJson(tx) {
  let parsed;
  try {
    parsed = JSON.parse(tx.toJson());
  } catch {
    return null;
  }
  const txV1 = parsed?.["1"];
  if (typeof txV1 !== "object" || txV1 === null) return null;
  const kind = txV1.kind;
  if (typeof kind !== "object" || kind === null) return null;
  if (kind.kind !== "programmable_transaction") return null;
  return kind;
}

function isPackageMakeImmutableCall(command) {
  return (
    typeof command === "object" &&
    command !== null &&
    command.command === "move_call" &&
    sameObjectId(command.package, FRAMEWORK_PACKAGE_ID) &&
    command.module === "package" &&
    command.function === "make_immutable"
  );
}

function inputMatchesObjectId(input, objectId) {
  return (
    typeof input === "object" &&
    input !== null &&
    ["immutable_or_owned", "receiving", "shared"].includes(input.type) &&
    sameObjectId(input.object_id, objectId)
  );
}

function publishesPackageAsImmutable(tx) {
  const programmable = programmableTransactionJson(tx);
  if (programmable === null) return false;
  const commands = programmable.commands;
  if (!Array.isArray(commands)) return false;
  const publishIndexes = commands
    .map((c, i) => (typeof c === "object" && c?.command === "publish" ? i : -1))
    .filter((i) => i !== -1);
  if (publishIndexes.length !== 1) return false;
  const publishIndex = publishIndexes[0];
  for (const command of commands.slice(publishIndex + 1)) {
    if (!isPackageMakeImmutableCall(command)) continue;
    const args = command.arguments;
    if (
      Array.isArray(args) &&
      args.length === 1 &&
      typeof args[0] === "object" &&
      args[0]?.result === publishIndex
    ) {
      return true;
    }
  }
  return false;
}

function usesUpgradeCapForMakeImmutable(tx, upgradeCapId) {
  const programmable = programmableTransactionJson(tx);
  if (programmable === null) return false;
  const inputs = programmable.inputs;
  const commands = programmable.commands;
  if (!Array.isArray(inputs) || !Array.isArray(commands)) return false;
  const upgradeCapInputs = inputs
    .map((input, i) =>
      inputMatchesObjectId(input, upgradeCapId.toHex()) ? i : -1,
    )
    .filter((i) => i !== -1);
  if (upgradeCapInputs.length === 0) return false;
  for (const command of commands) {
    if (!isPackageMakeImmutableCall(command)) continue;
    const args = command.arguments;
    if (!Array.isArray(args) || args.length !== 1) continue;
    const arg = args[0];
    if (
      typeof arg === "object" &&
      arg !== null &&
      typeof arg.input === "number" &&
      upgradeCapInputs.includes(arg.input)
    ) {
      return true;
    }
  }
  return false;
}

async function wasPackagePublishedAsImmutable(client, packageId) {
  let cursor = undefined;
  while (true) {
    const page = await client.transactionsDataEffects(
      TransactionsFilter.new({ changedObject: packageId }),
      forwardPage(cursor),
    );
    for (const txData of page.data) {
      if (publishesPackageAsImmutable(txData.signedTransaction.transaction))
        return true;
    }
    if (page.pageInfo.hasNextPage) cursor = page.pageInfo.endCursor;
    else return false;
  }
}

async function wasUpgradeCapUsedForMakeImmutable(client, upgradeCapId) {
  let cursor = undefined;
  while (true) {
    const page = await client.transactionsDataEffects(
      TransactionsFilter.new({ inputObject: upgradeCapId }),
      forwardPage(cursor),
    );
    for (const txData of page.data) {
      if (
        usesUpgradeCapForMakeImmutable(
          txData.signedTransaction.transaction,
          upgradeCapId,
        )
      )
        return true;
    }
    if (page.pageInfo.hasNextPage) cursor = page.pageInfo.endCursor;
    else return false;
  }
}

async function currentPackagePolicy(client, packageId) {
  const upgradeCapId = await resolveUpgradeCapId(client, packageId);
  if (upgradeCapId === null) {
    if (await wasPackagePublishedAsImmutable(client, packageId))
      return "Immutable";
    return "Unavailable";
  }
  const contents = await client.moveObjectContents(upgradeCapId);
  if (contents === null) {
    if (await wasUpgradeCapUsedForMakeImmutable(client, upgradeCapId))
      return "Immutable";
    return "Unavailable";
  }
  const policy = extractPolicy(contents);
  return policy !== null ? formatPolicyName(policy) : "Unavailable";
}

const packageId =
  "0x6f727ea576a00036657fff0ae3a6d7c8171b178bf35112d6b83b2a6272cc5f0d";
const packageAddress = Address.fromHex(packageId);
const client = GraphQlClient.newTestnet();

// Fetch package metadata and version history.
const pkg = await client.package_(packageAddress);
if (pkg === null) throw new Error("missing package");

const latestPackage = await client.packageLatest(packageAddress);
if (latestPackage === null) throw new Error("missing latest package");

const versions = await fetchPackageVersions(client, packageAddress);
const packagePrefix = pkg.id().toHex();
console.log(
  `Latest version: ${latestPackage.version().asU64()} (${latestPackage.id().toHex()})`,
);
// Resolve the current upgrade policy.
console.log(
  `Current package policy: ${await currentPackagePolicy(client, pkg.id())}\n`,
);

// Print the package version history.
console.log("Versions:");
for (const version of versions) {
  const labels = [];
  if (version.id().eq?.(pkg.id()) ?? false) labels.push("requested");
  if (version.id().eq?.(latestPackage.id()) ?? false) labels.push("latest");
  let line = `- v${version.version().asU64()} -> ${version.id().toHex()}`;
  if (labels.length > 0) line += ` [${labels.join(", ")}]`;
  console.log(line);
}
console.log();

// Print package dependencies and their linked versions.
console.log("Dependencies:");
const linkageTable = pkg.linkageTable();
if (linkageTable.size === 0) {
  console.log("- none");
} else {
  const upgrades = [...linkageTable.values()].sort((a, b) =>
    a.upgradedId.toHex() < b.upgradedId.toHex() ? -1 : 1,
  );
  for (const upgrade of upgrades) {
    console.log(
      `- ${upgrade.upgradedId.toHex()} @ v${upgrade.upgradedVersion.asU64()}`,
    );
  }
}
console.log();

// Inspect normalized modules, functions, types, and sample key objects.
console.log("Package contents:");
const moduleNames = [...pkg.modules().keys()].map((m) => m.asStr()).sort();

for (const moduleName of moduleNames) {
  console.log(`Module: ${moduleName}`);
  const module = await client.normalizedMoveModule(
    packageAddress,
    moduleName,
    undefined,
    forwardPage(),
    forwardPage(),
    forwardPage(),
    forwardPage(),
  );
  if (module === null) {
    console.log("  metadata: missing\n");
    continue;
  }
  if (module.functions === null || module.functions.nodes.length === 0) {
    console.log("  functions: none");
  } else {
    console.log("  functions:");
    for (const fun of module.functions.nodes) {
      console.log(
        `    - ${formatFunctionSignature(String(fun), packagePrefix)}`,
      );
    }
    if (module.functions.pageInfo.hasNextPage) console.log("    - ...");
  }
  if (module.structs === null || module.structs.nodes.length === 0) {
    console.log("  types: none");
  } else {
    console.log("  types:");
    for (const struct_ of module.structs.nodes) {
      const typeTag = `${packagePrefix}::${moduleName}::${struct_.name}`;
      console.log(`    - ${typeTag}`);
      const hasKeyAbility =
        struct_.abilities !== null &&
        struct_.abilities.includes(MoveAbility.Key);
      const isGeneric =
        struct_.typeParameters !== null && struct_.typeParameters.length > 0;
      await printObjectSamples(client, typeTag, hasKeyAbility, isGeneric);
    }
    if (module.structs.pageInfo.hasNextPage) console.log("    - ...");
  }
  console.log();
}
